File: LeetCode/114/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  void solve(TreeNode *node, TreeNode *&head) {
    if (node == nullptr)
      return;

    head->right = node;
    head->left = nullptr;

    TreeNode *left = node->left;
    TreeNode *right = node->right;

    head = head->right;

    solve(left, head);
    solve(right, head);
  }

  void flatten(TreeNode *root) {
    TreeNode dummy{0, nullptr, nullptr};
    TreeNode *head = &dummy;
    solve(root, head);
    root = head;
  }
};

File: LeetCode/114/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func solve(node *TreeNode, head **TreeNode) {
	if node == nil {
		return
	}

	(*head).Right = node
	(*head).Left = nil

	left, right := node.Left, node.Right

	(*head) = (*head).Right

	solve(left, head)
	solve(right, head)
}

func flatten(root *TreeNode) {
	dummy := &TreeNode{}
	head := dummy
	solve(root, &head)
}
